File: api/button-click.js
const {
  getRequestBody,
  jsonResponse,
  postToDiscord,
} = require("./_lib/discord");

module.exports = async function buttonClick(req, res) {
  if (req.method !== "POST") {
    res.setHeader("Allow", "POST");
    return jsonResponse(res, 405, { ok: false, error: "Method not allowed." });
  }

  const data = getRequestBody(req);
  const label = String(data.label == null ? "(unknown)" : data.label).slice(0, 200);
  const webhookUrl = (process.env.DISCORD_BUTTON_CLICK_WEBHOOK_URL || "").trim();

  if (!webhookUrl) {
    return jsonResponse(res, 503, {
      ok: false,
      error: "Webhook not configured.",
    });
  }

  const payload = {
    username: "Button Click Tracker",
    embeds: [
      {
        title: "Button clicked",
        color: 5814783,
        fields: [
          {
            name: "Button text",
            value: label || "(empty)",
            inline: false,
          },
        ],
        timestamp: new Date().toISOString(),
      },
    ],
  };

  try {
    const result = await postToDiscord(
      webhookUrl,
      payload,
      "ButtonClickNotifier/1.0",
    );

    if (!result.ok) {
      return jsonResponse(res, 502, {
        ok: false,
        error: `Discord rejected (HTTP ${result.status}).`,
      });
    }
  } catch {
    return jsonResponse(res, 502, {
      ok: false,
      error: "Could not reach Discord.",
    });
  }

  return jsonResponse(res, 200, { ok: true });
};

File: api/button-click.py
"""Vercel API route for non-sensitive button-click notifications."""

from __future__ import annotations

import json
import time
from http.server import BaseHTTPRequestHandler

from lib.discord import (
    json_response,
    method_not_allowed,
    post_webhook,
    read_body,
    webhook_url,
)


class handler(BaseHTTPRequestHandler):
    def do_POST(self) -> None:
        try:
            data = json.loads(read_body(self).decode("utf-8", errors="replace") or "{}")
            label = str(data.get("label", "(unknown)"))[:200]
        except (ValueError, TypeError, AttributeError):
            label = "(unknown)"

        target = webhook_url("DISCORD_BUTTON_CLICK_WEBHOOK_URL")
        if not target:
            json_response(self, 503, {"ok": False, "error": "Webhook not configured."})
            return

        payload = {
            "username": "Button Click Tracker",
            "embeds": [
                {
                    "title": "Button clicked",
                    "color": 5814783,
                    "fields": [
                        {"name": "Button text", "value": label or "(empty)", "inline": False},
                    ],
                    "timestamp": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
                }
            ],
        }
        status, error = post_webhook(target, payload, "ButtonClickNotifier/1.0")
        if error:
            json_response(
                self,
                502,
                {"ok": False, "error": "Could not reach Discord." if status is None else error},
            )
            return

        json_response(self, 200, {"ok": True})

    def do_GET(self) -> None:
        method_not_allowed(self)

    def do_PUT(self) -> None:
        method_not_allowed(self)

    def do_PATCH(self) -> None:
        method_not_allowed(self)

    def do_DELETE(self) -> None:
        method_not_allowed(self)

    def log_message(self, format: str, *args: object) -> None:
        return